features/332-sso/api.md

```diff
@@ -0,0 +1,311 @@
+```
+
+### Example
+
+```typescript
+skygear.auth.loginOAuthProviderWithPopup("google").then(user => {
+  // login suceedeed
+});
+```
+
+## Web browser with Redirect
+
+The flow changes the location of the current window to show the provider authorization UI.
+
+### API
+
+```typescript
+interface RedirectOptions extends PopupOptions {
+  // If provided, when the flow finishes, the browser will be redirected
+  // to this URL. Default to window.location.href if not provided.
+  callbackURL?: string;
+}
+
+// Create a new user with the provider identity.
+function loginOAuthProviderWithRedirect(providerID: string, options?: RedirectOptions & LoginOptions): Promise<void>;
+// Link the provider identity with the current logged in user.
+function linkOAuthProviderWithRedirect(providerID: string, options?: RedirectOptions): Promise<void>;
+
+function getLoginRedirectResult(): Promise<User>;
+function getLinkRedirectResult(): Promise<User>;
+```
+
+### Example
+
+```typescript
+skygear.auth.loginOAuthProviderWithRedirect("google").catch(error => {
+  // If we reach here, something went wrong and window.location.href is unchanged.
+});
+
+// If the above promise did not reject, the page was redirected so the `then` handler
+// will not be called.
+
+// Later on when the flow finishes, the user will be redirected to `callbackURL`.
+
+// Call this to retrieve the result of the flow.
+skygear.auth.getLoginRedirectResult().then(user => {
+  // login suceedeed
+});
+```
+
+# External access token flow
+
+This is not a flow defined by OAuth 2.0. The main usecase is for native applications that have already acquired an access token, for example, via [Facebook Login for iOS](https://developers.facebook.com/docs/facebook-login/ios/).
+
+## Possible Vulnerability
+
+Suppose Alice grants access to Mallory, a malicious application, via OAuth 2.0. Mallory now has the bearer access token of Alice, which is considered a proof of identity. Mallory can now impersonate Alice and login our application with the identity of Alice.
+
+Since this API is vulnerable by design, it is disabled by default. The developer is strongly recommended to use the standard OAuth 2.0 Authorization Code flow even the application is native and has already had acquired an access token.
+
+### API
+
+```typescript
+function loginOAuthProviderWithAccessToken(providerID: string, accessToken: string, options?: LoginOptions): Promise<User>;
+function linkOAuthProviderWithAccessToken(providerID: string, accessToken: string): Promise<User>;
+```
+
+### Example
+
+```typescript
+getAccessTokenViaFacebookSDKSomehow().then(accessToken => {
+  loginOAuthProviderWithAccessToken("facebook", accessToken).then(user => {
+    // login suceedeed
+  });
+});
+```
+
+# Remove OAuth Provider
+
+If the removal will result in no login principals, then removal is forbidden. Such case happens when the user logged in with Google and has no other login means.
```

Comment:
As mentioned in #334, removal of current principal is disallowed.

Reply:
I will look into #334 which is still a PR now later.

However, we definitely need a way to make unlinking a OAuth login to a user account possible.